cmd/simulator/config/simulation.go

```diff
@@ -0,0 +1,184 @@
+package config
+
+import (
+	"encoding/json"
+	"fmt"
+	"math/big"
+
+	"github.com/smartcontractkit/libocr/offchainreporting2plus/types"
+)
+
+var (
+	ErrEncoding = fmt.Errorf("encoding/decoding failure")
+)
+
+// SimulationPlan is a collection of configurations with which to run a
+// simulation.
+type SimulationPlan struct {
+	Node            Node                  `json:"node"`
+	Network         Network               `json:"p2pNetwork"`
+	RPC             RPC                   `json:"rpc"`
+	Blocks          Blocks                `json:"blocks"`
+	ConfigEvents    []OCR3ConfigEvent     `json:"-"`
+	GenerateUpkeeps []GenerateUpkeepEvent `json:"-"`
+	LogEvents       []LogTriggerEvent     `json:"-"`
+}
+
+// Encode applies JSON encoding of a simulation plan to bytes.
+func (p SimulationPlan) Encode() ([]byte, error) {
+	type encodedOutput struct {
+		SimulationPlan
+		Events []interface{} `json:"events"`
+	}
+
+	encodable := encodedOutput{
+		SimulationPlan: p,
+		Events:         make([]interface{}, len(p.ConfigEvents)+len(p.GenerateUpkeeps)),
+	}
+
+	for _, event := range p.ConfigEvents {
+		// ensure the type is set properly
+		event.Type = OCR3ConfigEventType
+		encodable.Events = append(encodable.Events, event)
+	}
+
+	for _, event := range p.GenerateUpkeeps {
+		// ensure the type is set properly
+		event.Type = GenerateUpkeepEventType
+		encodable.Events = append(encodable.Events, event)
+	}
+
+	for _, event := range p.LogEvents {
+		// ensure the type is set properly
+		event.Type = LogTriggerEventType
+		encodable.Events = append(encodable.Events, event)
+	}
+
+	return json.Marshal(encodable)
+}
+
+// DecodeSimulationPlan uses JSON encoding to decode bytes to a simulation plan.
+func DecodeSimulationPlan(encoded []byte) (SimulationPlan, error) {
+	var plan SimulationPlan
+
+	if err := json.Unmarshal(encoded, &plan); err != nil {
+		return plan, fmt.Errorf("%w: failed to decode simulation plan: %s", ErrEncoding, err.Error())
+	}
+
+	plan.ConfigEvents = make([]OCR3ConfigEvent, 0)
+	plan.GenerateUpkeeps = make([]GenerateUpkeepEvent, 0)
+	plan.LogEvents = make([]LogTriggerEvent, 0)
+
+	type eventCollection struct {
```

Comment:
Can't we have a specific field for each event type, and avoid the "manual" decoding?

Reply:
The events are in a single JSON array and each event type has a `type` property for decoding. This JSON structure allows a person to visually order events by block number instead of having to jump between properties. All events can be viewed as if on a linear timeline.

```
"events": [
  ... event 1
  ... event 2
  ...
]
```

instead of the following each with their own ordering

```
"conditionalUpkeeps": [],
"logTriggerUpkeeps": [],
"ocrConfig": [],
"logTriggers": []
```

The intention was to make the config JSON more manageable with the consequence of adding more complexity to the unmarshaling.

Reply:
In that case, we can have a single struct for event, with pointer fields that can be nil depends on the actual type.

Reply:
I don't understand. Is your proposal to collapse all fields for all events into a single 'god object'? Right now, all events share a basic event structure:

```
type Event struct {
	Type         EventType `json:"type"`
	TriggerBlock *big.Int  `json:"eventBlockNumber"`
	Comment      string    `json:"comment,omitempty"`
}
```

And then each specific event extends the basic event like the following:

```
type SomeEvent struct {
	Event
	SomeSpecificProperty string `json:"someSpecificProperty"`
}
```

This method provides more strict types within Go with the consequence being a little more complexity in json decoding.

Reply:
You can have a dedicated field for each extension, e.g.
```
type Event struct {
	Type         EventType `json:"type"`
	TriggerBlock *big.Int  `json:"eventBlockNumber"`
	Comment      string    `json:"comment,omitempty"`
	SomeEvent *SomeEvent `json:"someEvent,omitempty"`
}
```

This is a common pattern in golang because the lack of good inheritance (for example this is the way to go when you use protobuf).

Reply:
So then the JSON would look something like the following?

```
{
  "type": "theType",
  "eventBlockNumber": 10,
  "someEvent": {
    "someProp": 420
  }
}
```

How does this help? The simulation plan is already divided into distinct types and `Event` is never directly used. It only exists for decoding. There is a switch statement that in `tools/simulator/config/simulation.go` starting at line `88` that organizes the simulation events based on encoded type. This would still exist for your method to sort after decoding. The only difference would be a single `json.Unmarshal` instead of multiple for each event. Plus validation would need to exist such that a person should not be able to do the following:

```
{
  "type": "theType",
  "eventBlockNumber": 10,
  "someEvent": {
    "someProp": 420
  },
  "otherEvent": {
    "eventProp": "value"
  }
}
```

Reply:
I think I prefer having separate structs for each event, rather than having a single struct that acts as a catch-all for all events? I think it's just cleaner/more focused having a single struct per event type rather than  a single struct that you either have to have separate fields/objects per event OR if you try to "flatten" all the distinct event fields into the top level, it can get tricky trying to identify what type of event an instance of the god struct is representing, especially if multiple events have field names in common etc

Reply:
The Go-ish way is NOT to have such inheritance, the result is a messy & manual code for encoding/decoding.
You can find examples for this all over with protobufs, and also in our code we have `Trigger` which holds `LogTriggerExtension` which is exactly what I'm suggesting.

Protecting the user from doing mistakes can help in cases we have some external interface that might change over time, but in this case the users are internal (and technical).

We can go with this pattern JUST for the simulator, but this is something we should avoid doing in our operational code.